File: JavaScript/algorithm/src/algorithm/혼자놀기의달인/혼자놀기의달인.js
function solution(cards) {
  let answer = 0;

  let cardIndex = 0;
  const openBox = Array.from({ length: cards.length }, (el, key) =>
    key === 0 ? true : false
  );

  const arr = [8];
  const table = new Map();
  let tableIndex = 1;

  for (let i = 0; i < cards.length; ++i) {
    const num = cards[cardIndex] - 1; // 다음 박스 번호
    console.log("num : ", num);
    cardIndex = num;

    if (!openBox[num]) {
      table.set(tableIndex, (table.get(tableIndex) || 0) + 1);
      arr.push(cards[num]);
      console.log("arr : ", arr);

      console.log("openBox[num] : ", openBox[num]);

      openBox[num] = true;
    } else {
      tableIndex++;
      openBox[num] = true;
      // i--;

      for (let j = 1; j < openBox.length; ++j) {
        if (!openBox[j]) {
          cardIndex = j;
          break;
        }
      }
    }
  }
  console.log("table : ", table);
  console.log("openBox : ", openBox);

  return answer;
}

console.log(solution([8, 6, 3, 7, 2, 5, 1, 4]));
// 12

// https://school.programmers.co.kr/learn/courses/30/lessons/131130

File: python/algorithm/everything_algorithm/algorithm_study/혼자놀기의달인/혼자놀기의달인.py
def solution(cards):
    answer = 0

    openBox = []
    for i in range(len(cards)):
        if (i == 0): openBox.append(True)
        else: openBox.append(False)


    table = {}
    tableIndex = 1
    cardIndex = 0
    table[tableIndex] = 1


    while not all(openBox):
        num = cards[cardIndex] - 1
        cardIndex = num

        if (not openBox[num]):
            if tableIndex in table:
                table[tableIndex] = table[tableIndex] + 1
            else:
                table[tableIndex] = 1
        else:
            tableIndex = tableIndex + 1
            for i in range(1, len(openBox)):
                if not openBox[i]:
                    cardIndex = i
                    break

        openBox[num] = True

    arr = []
    for key, value in table.items():
        arr.append([key, value])

    for i in range(len(arr) - 1):
        for j in range(i + 1, len(arr)):
            answer = max(answer, arr[i][1] * arr[j][1])

    return answer


print(solution([8, 6, 3, 7, 2, 5, 1, 4]))
# 12